File: LeetCode/1011.Capacity_To_Ship_Packages_Within_D_Days/Mario/1011.Capacity_To_Ship_Packages_Within_D_Days.py
#!/usr/bin/env python3


class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        # break this weights into days segments, the sum of this segnment < res
        # get mini res
        n = len(weights)

        def check(x):
            cnt, seg = 1, 0
            for w in weights:
                if seg + w > x:
                    seg = 0
                    cnt += 1
                seg += w
            return cnt <= days

        left, right = max(weights), sum(weights)
        while left < right:
            mid = (left + right) >> 1
            if check(mid):
                right = mid
            else:
                left = mid + 1

        return left

File: LeetCode/1011.Capacity_To_Ship_Packages_Within_D_Days/1011.Capacity_To_Ship_Packages_Within_D_Days.swift
class Solution {
    func shipWithinDays(_ weights: [Int], _ days: Int) -> Int {
        let n = weights.count
        var l = weights.max()!, r = weights.reduce(0, +) + 1

        // can ship within days
        func canShip(_ w: Int) -> Bool {
            var res = 1, prev = 0
            for cur in weights {
                if cur + prev > w {
                    res += 1
                    prev = cur
                } else {
                    prev += cur
                }
            }
            return res <= days
        }

        var res = Int.max
        while l < r {
            let m = (l + r) / 2
            if canShip(m) {
                r = m
                res = min(res, m)
            } else {
                l = m + 1
            }
        }

        return res
    }
}
